README.md

```diff
@@ -120,13 +132,27 @@ Current state | Next word | Next state
 --------------|-----------|-----------
     S0        |    the    |     S1
     S1        |   . . .   |   . . .
+    
 
 (hint: my version has 13 entries. Yours _might_ be different)
 
 ## A2.4
 
-«replace this with your answer»
-
+Current state | Next word | Next state
+--------------|-----------|-----------
+    S0        |    the    |     S1
+    S1        |    lazy   |     S2
+    S1        |    smelly |     S2
+    S1        |    ""     |     S2
+    S1        |    lazy   |     S1
+    S1        |    smelly |     S1
+    S2        |    cat    |     S3
+    S2        |    dog    |     S3
+    S3        |    ran    |     S4
+    S3        |    ate    |     S4
+    S4        |    slowly |     eoi
+    S4        |    noisily|     eoi
+    S4        |    ""     |     eoi
 
 ## Q2.5 (12 points)
```

Comment:
Your code was on the correct path. You had a data structure representing the table. You just needed to change it slightly to allow it to be traversed. And if you had the correct state table to base it on, you'd find your code would become a lot easier.

Here's a possible version:

~~~ c++
#include <iostream>
using namespace std;
#include <map>

typedef int state;
typedef std::map<std::string, state> transitions;
typedef std::map<state, transitions> state_table;

typedef vector<string> sentence;


const int END = 99;

state_table states = {
// current    transitions
// state      {word -> new state}
// ---------+-------------------------------------------------
  {     0,    {{ "the",    1 }}},
  {     1,    {{ "lazy",   2 }, { "smelly",  2 },   { "dog", 3 }, {"cat", 3 }}},
  {     2,    {{ "lazy",   2 }, { "smelly",  2 },   { "dog", 3 }, {"cat", 3 }}},
  {     3,    {{ "ate",    4 }, { "ran",     4 }}},
  {     4,    {{ "slowly", 5 }, { "noisily", 5 },   { "EOI", END }}},
  {     5,    {{ "EOI",    END }}}
};


int getNextState(int currentState, string nextWord) {
	transitions t = states[currentState];
	if (t.count(nextWord))
		return t[nextWord];
	else
		return -1;
}

bool validSentence(sentence words) {
    int state = 0;
    for (string word: words) {
        state = getNextState(state, word);
        if (state < 0)
            return false;
    }
    return state == 5;
}
~~~

# -8